students/eowyn/session06/mailroom/mailroom.py

```diff
@@ -0,0 +1,237 @@
+    head = head + (width * "-") + '\n'
+    return head
+
+
+def setup_body(list_data):
+    ''' Set up the donor records table body '''
+    body = ""
+    fs2 = ''.join(["{:<40}", "${:^12.2f}", "{:^12d}", "${:^12.2f}"])
+    for i in range(len(list_data)):
+        entry = list_data[i]
+        body = body + fs2.format(*entry) + '\n'
+    return body
+
+
+def print_table(list_data):
+    ''' Pretty-print the donor records '''
+    print(setup_table())
+    print(setup_body(list_data))
+
+
+''' --------------------------------------------------------------
+Following are helper functions for adding donations to DONORS and
+sending thank you notes to donors
+-------------------------------------------------------------------'''
+
+
+def collect_donor_input():
+    ''' Get name and donation amount to add to DONORS'''
+    fullname = input("Enter a donor name (new or existing):\n")
+    fullname = remove_inputquotes(fullname)
+    amount = float(input("Donation amount: "))
+    return (fullname, amount)
+
+
+def retrieve_donations(fullname):
+    ''' if donor exists, return donations, otherwise, return None'''
+    if fullname in DONORS:
+        return DONORS[fullname]
+
+
+def add_donation(fullname, amount):
+    ''' Update donor if it exists or add new donor if it does not'''
+    DONORS.setdefault(fullname, []).append(amount)
+
+
+def update_donors():
+    '''
+    Add new donation to DONORS; if new donor, add to DONOR.
+    Then print a thank-you note for the donation.
+    '''
+    (fullname, amount) = collect_donor_input()
+    add_donation(fullname, amount)
+    print(generate_letter(fullname))
+
+
+def list_donors():
+    ''' List all DONORS '''
+    print("All Donors:")
+    [print(x) for x in DONORS]
+
+
+def generate_letter(donor_name):
+    '''
+    Generate a formatted thank you note to a donor and return
+    the string of the thank you note for printing or writing.
+    '''
+    fs = "Thank you, {}, for your generosity and recent gift of ${}."
+    return fs.format(donor_name, DONORS[donor_name][-1])
+
+
+''' --------------------------------------------------------------
+Following are helper functions for accepting and responding to
+keyboard input
+-------------------------------------------------------------------'''
+
+
+def remove_inputquotes(a_string):
+    '''Remove auxillary quotes to cleanse a_string'''
+    a_string.replace('"', '')
+    return a_string
+
+
+def get_user_input(prompt_string):
+    ''' Print a prompt_string, return keyboard input if no exceptions'''
+    try:
+        answer = input(prompt_string)
+    except (EOFError, KeyboardInterrupt, TypeError):
+        return None
+    else:
+        return answer
+
+
+def select_action(arg_dict, answer):
+    ''' Execute an action from arg_dict that corresponds to answer.
+    Return None if action was executed and False if an error occurs'''
+    try:
+        return arg_dict.get(answer)()
+    except (KeyError, TypeError):
```

Comment:
If you  get a TypeError here it means the arg_dict is broken -- do you want to catch that?